Doc(logging.md): add notes for logging using custom zap logger (#2599)
operator-framework/operator-sdk: doc/user/logging.md

	scopedLog.V(1).Info("Printing at DEBUG level")
}

> ```

Reviewer: Keep this line.

{"level":"info","ts":1559866310.065119,"logger":"scoped","msg":"Printing at INFO level"}
{"level":"debug","ts":1559866310.065123,"logger":"scoped","msg":"Printing at DEBUG level"}
```
## Custom zap logger

In order to use a custom zap logger, [`zapr`][repo_zapr] can be utilized to wrap it in a logr implementation.

Below is an example illustrating the use of [`zap-logfmt`][logfmt_repo] in logging. Here, the `NewLogger` function produces a new `logr.Logger` from an existing `zap.Logger`.

### Example

```Go
package main

import (
	"github.com/go-logr/zapr"
	zaplogfmt "github.com/sykesm/zap-logfmt"
	"go.uber.org/zap/zapcore"
	logf "sigs.k8s.io/controller-runtime/pkg/log"
)

var log logr.Logger

func main() {
	configLog := zap.NewProductionEncoderConfig()
	configLog.EncodeTime = func(ts time.Time, encoder zapcore.PrimitiveArrayEncoder) {
		encoder.AppendString(ts.UTC().Format(time.RFC3339))
	}

	logger := zap.New(zapcore.NewCore(
		zaplogfmt.NewEncoder(configLog),
		os.Stdout,
		zapcore.DebugLevel,
	))

	// Construct a new logr.logger. 
	log = zapr.NewLogger(logger)

> 	// Instantiate the logger. This logger will

Reviewer: ```suggestion
	// Set the controller logger to log, which will
```

{"level":"info","ts":1559866310.065119,"logger":"scoped","msg":"Printing at INFO level"}
{"level":"debug","ts":1559866310.065123,"logger":"scoped","msg":"Printing at DEBUG level"}
```
## Custom zap logger

In order to use a custom zap logger, [`zap`][controller_runtime_zap] from controller-runtime can be utilized to wrap it in a logr implementation.

Below is an example illustrating the use of [`zap-logfmt`][logfmt_repo] in logging.

### Example

```Go
package main

import (

> 	"github.com/go-logr/zapr"

Reviewer 2: IMO: 
- here we need to add just the imports used required for this impl in the main.go file scaffolded by default. 
- Also, we to use ... / ... as the following example to illustrate and make clear that it will have other imports. 

```go
import (
        ... 
	"github.com/go-logr/zapr"
	zaplogfmt "github.com/sykesm/zap-logfmt"
	"go.uber.org/zap/zapcore"
	"go.uber.org/zap"
	logf "sigs.k8s.io/controller-runtime/pkg/log"
	zapcr "sigs.k8s.io/controller-runtime/pkg/log/zap"
         ...

```

> 	zaplogfmt "github.com/sykesm/zap-logfmt"

Reviewer 2: Why we need the `zaplogfmt "github.com/sykesm/zap-logfmt"`? Should we not use what the controller-runtime offers?

	"go.uber.org/zap/zapcore"
	"go.uber.org/zap"
	logf "sigs.k8s.io/controller-runtime/pkg/log"
	zapcr "sigs.k8s.io/controller-runtime/pkg/log/zap"
)

var log logr.Logger

> func main() {

Reviewer 2: IMO:
- Currently, we scaffold in the project an implementation to do the logs. We need to make clear that X should be replaced by Y or kept. 
- Also, the doc need to explain how to do it in the project built with SDK and not give a generic example.

	configLog := zap.NewProductionEncoderConfig()

> 	configLog.EncodeTime = func(ts time.Time, encoder zapcore.PrimitiveArrayEncoder) {

Reviewer 2: IHMO It could be something such as how to:

**Example**

In your main.go replace the following implementation;

```go
func main() {
        ....
	// Add the zap logger flag set to the CLI. The flag set must
	// be added before calling pflag.Parse().
	pflag.CommandLine.AddFlagSet(zap.FlagSet())

	// Add flags registered by imported packages (e.g. glog and
	// controller-runtime)
	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)

	pflag.Parse()

	// Use a zap logr.Logger implementation. If none of the zap
	// flags are configured (or if the zap flag set is not being
	// used), this defaults to a production zap logger.
	//
	// The logger instantiated here can be changed to any logger
	// implementing the logr.Logger interface. This logger will
	// be propagated through the whole operator, generating
	// uniform and structured logs.
	logf.SetLogger(zap.Logger())
```

With:

```go
	configLog := zap.NewProductionEncoderConfig()
	configLog.EncodeTime = func(ts time.Time, encoder zapcore.PrimitiveArrayEncoder) {
		encoder.AppendString(ts.UTC().Format(time.RFC3339))
	}
	logfmtEncoder := zaplogfmt.NewEncoder(configLog)

	// Construct a new logr.logger.
	log = zapcr.New(zapcr.UseDevMode(true), zapcr.WriteTo(os.Stdout), zapcr.Encoder(logfmtEncoder))

	// Set the controller logger to log, which will
	// be propagated through the whole operator, generating
	// uniform and structured logs.
	logf.SetLogger(log)
```

And then, ensure that you are using the following imports:

```go 
imports(
         ...
	"go.uber.org/zap/zapcore"
	"go.uber.org/zap"
	logf "sigs.k8s.io/controller-runtime/pkg/log"
	zapcr "sigs.k8s.io/controller-runtime/pkg/log/zap"
        zaplogfmt "github.com/sykesm/zap-logfmt"
        ...
)
```

**NOTE:** You will need to add the module `"github.com/sykesm/zap-logfmt"` to your project. Run `go get -u github.com/sykesm/zap-logfmt`

To test you can, for example, add the following print in the main function: 

```go
log.Info("Printing at DEBUG LEVEL")
```

So, when you run your operator you will check the following result. 

```
$ operator-sdk run --local 
INFO[0000] Running the operator locally in namespace default. 
ts=2020-02-27T20:03:14Z level=info msg="Printing at DEBUG LEVEL"
ts=2020-02-27T20:03:14Z level=info msg="Operator Version: 0.0.1"
ts=2020-02-27T20:03:14Z level=info msg="Go Version: go1.13.3"
ts=2020-02-27T20:03:14Z level=info msg="Go OS/Arch: darwin/amd64"
ts=2020-02-27T20:03:14Z level=info msg="Version of operator-sdk: v0.15.0+git"

```

Reviewer 2: However, note that I perform these changes and the level still info. Should not it be debug?
@hasbro17 @estroz

Reviewer 2: I did the changes and the level is info. Is it correct? Should not be debug?

		encoder.AppendString(ts.UTC().Format(time.RFC3339))
	}
	logfmtEncoder := zaplogfmt.NewEncoder(configLog)

	// Construct a new logr.logger.
	log = zapcr.New(zapcr.UseDevMode(true), zapcr.WriteTo(os.Stdout), zapcr.Encoder(logfmtEncoder))

	// Set the controller logger to log, which will
	// be propagated through the whole operator, generating
	// uniform and structured logs.
	logf.SetLogger(log)

	log.Info("Printing at DEBUG LEVEL")	
}
```

#### Output using custom zap logger
```console
$ go run main.go
ts=2020-02-25T22:42:43Z level=info msg="Printing at DEBUG level"

> ts=2020-02-25T22:42:43Z level=info msg="Printing at DEBUG level"

Reviewer 2: The output needs have been obtained from the Memcached Go Sample with these changes. 
We need to show how the logs in the operator project container will be. 
See: https://github.com/operator-framework/operator-sdk-samples/tree/master/go/memcached-operator

{"level":"info","ts":1559866310.065119,"logger":"scoped","msg":"Printing at INFO level"}
{"level":"debug","ts":1559866310.065123,"logger":"scoped","msg":"Printing at DEBUG level"}
```
## Custom zap logger

In order to use a custom zap logger, [`zapr`][repo_zapr] can be utilized to wrap it in a logr implementation.

Below is an example illustrating the use of [`zap-logfmt`][logfmt_repo] in logging. Here, the `NewLogger` function produces a new `logr.Logger` from an existing `zap.Logger`.

### Example

```Go
package main

import (
	"github.com/go-logr/zapr"
	zaplogfmt "github.com/sykesm/zap-logfmt"
	"go.uber.org/zap/zapcore"
	logf "sigs.k8s.io/controller-runtime/pkg/log"
)

var log logr.Logger

func main() {

> 	configLog := zap.NewProductionEncoderConfig()

Reviewer 3: Where is the `zap` import defined?

	configLog.EncodeTime = func(ts time.Time, encoder zapcore.PrimitiveArrayEncoder) {
		encoder.AppendString(ts.UTC().Format(time.RFC3339))
	}

	logger := zap.New(zapcore.NewCore(
		zaplogfmt.NewEncoder(configLog),
		os.Stdout,
		zapcore.DebugLevel,

> 	))

Reviewer 2: Should not it be something such as following?

```suggestion
	logger := zap.New(zap.UseDevMode(true))
```
See kubebuider: https://github.com/kubernetes-sigs/kubebuilder/pull/1385/files
c/c @hasbro17

Author: I have followed an implementation of zaplogfmt from here - https://godoc.org/github.com/sykesm/zap-logfmt#example-package--Usage

Reviewer 3: Right, so since we're not really able to use the SDK's [`pkg/log/zap`](https://github.com/operator-framework/operator-sdk/blob/master/pkg/log/zap/logger.go) for this example to instantiate a new zapr based logger (since that doesn't let you override the encoder or the format), we might as well use the controller-runtime's [`pkg/log/zap`](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/log/zap/zap.go) pkg and use its [zap.New()](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/log/zap/zap.go#L35) to instantiate a zapr based logger and pass in our own zaplogfmt  encoder.
That should be a little cleaner with functional options and more in line with upstream conventions.

So @varshaprasad96 I think this would make the example something like this:
```Go
import (
	"github.com/go-logr/zapr"
	zaplogfmt "github.com/sykesm/zap-logfmt"
	"go.uber.org/zap/zapcore"
	"go.uber.org/zap"
	logf "sigs.k8s.io/controller-runtime/pkg/log"
        zapcr "sigs.k8s.io/controller-runtime/pkg/log/zap"
)

var log logr.Logger

func main() {
	configLog := zap.NewProductionEncoderConfig()
	configLog.EncodeTime = func(ts time.Time, encoder zapcore.PrimitiveArrayEncoder) {
		encoder.AppendString(ts.UTC().Format(time.RFC3339))
	}
	logfmtEncoder := zaplogfmt.NewEncoder(configLog)
	logger := zapcr.New(zapcr.UseDevMode(true), zapcr.WriteTo(os.Stdout), zapcr.Encoder(logfmtEncoder))

...

}
```

Author: @hasbro17 @camilamacedo86 That makes sense. Thanks! I have made the necessary changes. Please have a look.

	// Construct a new logr.logger. 
	log = zapr.NewLogger(logger)

	// Set the controller logger to log, which will
	// be propagated through the whole operator, generating
	// uniform and structured logs.

> 	logf.SetLogger(log)

Reviewer 2: Cannot bee simplified as `

logf.SetLogger(zap.New(zap.UseDevMode(true)))`

}
```

#### Output using custom zap logger
```console
$ go run main.go
ts=2020-02-25T22:42:43Z level=info msg="Printing at INFO level"

> ts=2020-02-25T22:42:43Z level=info msg="Printing at INFO level"

Reviewer 2: The logs shows INFO level, but the example is adding `zapcore.DebugLevel`. So, is this logs from the above example? Should not it bee debug level?

{"level":"info","ts":1559866310.065119,"logger":"scoped","msg":"Printing at INFO level"}
{"level":"debug","ts":1559866310.065123,"logger":"scoped","msg":"Printing at DEBUG level"}
```
## Custom zap logger

In order to use a custom zap logger, [`zap`][controller_runtime_zap] from controller-runtime can be utilized to wrap it in a logr implementation.

Below is an example illustrating the use of [`zap-logfmt`][logfmt_repo] in logging.

### Example

> In your main.go file, replace the following implementation inside the main function:

Reviewer 2: ```suggestion
In your `main.go file`, replace the current implementation for logs inside the `main` function :
```

> In your cmd/manager.main.go file, replace the following implementation inside the main function:

Reviewer 2: ```suggestion
In your `cmd/manager.main.go` file, replace the following implementation inside the main function:
```

```Go
	...
	// Add the zap logger flag set to the CLI. The flag set must
	// be added before calling pflag.Parse().
	pflag.CommandLine.AddFlagSet(zap.FlagSet())

	// Add flags registered by imported packages (e.g. glog and
	// controller-runtime)
	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)

	pflag.Parse()

	// Use a zap logr.Logger implementation. If none of the zap
	// flags are configured (or if the zap flag set is not being
	// used), this defaults to a production zap logger.
	//
	// The logger instantiated here can be changed to any logger
	// implementing the logr.Logger interface. This logger will
	// be propagated through the whole operator, generating
	// uniform and structured logs.
	logf.SetLogger(zap.Logger())

> 	...

Reviewer 2: The indentation needs are equals other docs. See: https://github.com/operator-framework/operator-sdk/blob/master/doc/user-guide.md#reconcile-loop

```

With:

```Go
	configLog := zap.NewProductionEncoderConfig()
	configLog.EncodeTime = func(ts time.Time, encoder zapcore.PrimitiveArrayEncoder) {
		encoder.AppendString(ts.UTC().Format(time.RFC3339))
	}
	logfmtEncoder := zaplogfmt.NewEncoder(configLog)

	// Construct a new logr.logger.
	log = zapcr.New(zapcr.UseDevMode(true), zapcr.WriteTo(os.Stdout), zapcr.Encoder(logfmtEncoder))

	// Set the controller logger to log, which will
	// be propagated through the whole operator, generating
	// uniform and structured logs.
	logf.SetLogger(log)
```

Ensure that the following additional imports are being used:

```
imports(
		...
		zaplogfmt "github.com/sykesm/zap-logfmt"
		zapcr "sigs.k8s.io/controller-runtime/pkg/log/zap"
		logf "sigs.k8s.io/controller-runtime/pkg/log"
		"go.uber.org/zap"
		"go.uber.org/zap/zapcore"
		...

> )

Reviewer 2: dito^

>         ...

Reviewer 2: ^fix the indentation.

```
**NOTE**: For this example, you will need to add the module `"github.com/sykesm/zap-logfmt"` to your project. Run `go get -u github.com/sykesm/zap-logfmt`.

To test, the following print statement can be added in the main function:

> `log.Info("Printing at DEBUG LEVEL")`

Reviewer 2: May, we should not use the DEBUG LEVEL word, by running this example the level will be == info

Author: Will change this to info.  We arent setting it to debug level in the example.

#### Output using custom zap logger

> ```console

Reviewer 2: Missing space

```suggestion

```console
```

$ operator-sdk run --local

> ts=2020-02-27T23:10:33Z level=info msg="Prinitng at DEBUG level"

Reviewer 2: ```suggestion
ts=2020-02-27T23:10:33Z level=info msg="Printing at INFO LEVEL"
```

Reviewer 2: I mean just that ^

> ts=2020-02-27T23:10:33Z level=info msg="Prinitng at DEBUG level"

Reviewer 2: Then we need change here too.

Author: This is the output of running locally too. Ill change the command.

ts=2020-02-27T23:10:33Z level=info msg="Operator Version: 0.0.1"
ts=2020-02-27T23:10:33Z level=info msg="Go Version: go1.13.8"
ts=2020-02-27T23:10:33Z level=info msg="Go OS/Arch: darwin/amd64"

> ts=2020-02-27T23:10:33Z level=info msg="Version of operator-sdk: v0.15.2"

Reviewer 2: Here IMO we need to show the result of the operator running. 
To run locally we use `operator-sdk run --local`

[code_memcached_controller]:../../example/memcached-operator/memcached_controller.go.tmpl
[code_set_logger]:https://github.com/operator-framework/operator-sdk/blob/4d66be409a69d169aaa29d470242a1defbaf08bb/internal/pkg/scaffold/cmd.go#L92-L96
[zap_sampling]:https://github.com/uber-go/zap/blob/master/FAQ.md#why-sample-application-logs

> [logfmt_repo]: https://github.com/jsternberg/zap-logfmt

Reviewer: ```suggestion
[logfmt_repo]:https://github.com/jsternberg/zap-logfmt
```
